memory/StackMap.h

```diff
@@ -0,0 +1,302 @@
+#pragma once
+#include "utils/StdExtensions.h"
+#include <array>
+#include <cinttypes>
+#include <cstring>
+
+namespace memory
+{
+
+template <typename KeyT, typename T, uint32_t SIZE>
+class StackMap
+{
+    struct ElementEntry
+    {
+        std::pair<KeyT, T> keyValue;
+        bool committed = false;
+    };
+
+    struct IndexEntry
+    {
+        IndexEntry() : keyHash(0), position(0) {}
+        IndexEntry(uint64_t key, uint32_t position) : keyHash(key & 0xFFFFFFFFF), position(position) {}
+
+        uint64_t keyHash;
+        uint32_t position;
+    };
+
+public:
+    class IterBase
+    {
+    public:
+        typedef std::pair<KeyT, T> value_type;
+        IterBase(ElementEntry* entries, uint32_t pos, uint32_t endPos) : _elements(entries), _pos(pos), _end(endPos) {}
+        IterBase(const IterBase& it) : _elements(it._elements), _pos(it._pos), _end(it._end) {}
+
+        IterBase& operator++()
+        {
+            if (_pos == _end)
+            {
+                return *this; // cannot advance a logical end iterator
+            }
+
+            ++_pos;
+            while (_pos != _end && !_elements[_pos].committed)
+            {
+                ++_pos;
+            }
+
+            return *this;
+        }
+
+        value_type& operator*() { return _elements[_pos].keyValue; }
+        value_type* operator->() { return &_elements[_pos].keyValue; }
+        const value_type& operator*() const { return _elements[_pos].keyValue; }
+        const value_type* operator->() const { return &_elements[_pos].keyValue; }
+        bool operator==(const IterBase& it) const { return _pos == it._pos || (isEnd() && it.isEnd()); }
+
+        bool operator!=(const IterBase& it) const
+        {
+            if (!isEnd() && !it.isEnd())
+            {
+                return _pos != it._pos;
+            }
+            return isEnd() != it.isEnd();
+        }
+
+    private:
+        bool isEnd() const { return _pos == _end; }
+        ElementEntry* _elements;
+        uint32_t _pos;
+        uint32_t _end;
+    };
+
+    typedef const IterBase const_iterator;
+    typedef IterBase iterator;
+    typedef std::pair<KeyT, T> value_type;
+
+    explicit StackMap() : _end(SIZE), _maxSpread(1), _count(0) {}
+
+    std::pair<iterator, bool> add(const KeyT& key, const T& value)
+    {
+        auto keyHash = utils::hash<KeyT>{}(key);
+
+        for (uint32_t i = 0; i < _maxSpread; ++i)
+        {
+            const auto& indexEntry = _index[indexPosition(keyHash, i)];
+            if (indexEntry.keyHash == keyHash && indexEntry.position)
+            {
+                return std::make_pair(end(), false); // no duplets
```

Comment:
This will prevent add elements with hash collisions as well. But there is a simple way to solve that that (we will need twice the memory to `_index`and rerange that after an an X number of erases, but we could augment the load factor to reduce the memory, 25% sounds a bit exaggerated)

Reply:
I know, that is the weakness of this simple solution. We could store the original key, compare on that and decide to find another slot to the right and increase _maxSpread.

Reply:
It is the same with MpmcHashMap. The key is limited to 40 bits in that case and it is necessary to allow atomic read and update of the key and position. But it means a hash collision prevents you from adding that object.
Comparing the original keys would render a sequence of atomic reads, and double reads that makes it impractical to use. And much harder to implement.

Reply:
In this case with 64 bits, hash collision is very rare unless the hash function is poor.